source/agora/node/Ledger.d

```diff
@@ -25,17 +26,21 @@ import agora.node.API;
 import vibe.core.log;
 
 import std.algorithm;
+import std.file;
+import std.path;
 
 /// Ditto
 public class Ledger
 {
-    /// data storage for all the blocks,
+    /// Blockstorage for all the blocks,
     /// currently a single contiguous region to
-    /// improve locality of reference
-    private const(Block)[] ledger;
+    private BlockStorage ledger;
 
-    /// pointer to the latest block
-    private const(Block)* last_block;
+    /// latest block of `BlockStorage`
+    private Block last_block;
+
+    /// is vaild `last_block`
+    private bool is_valid_last_block;
```

Comment:
This is unnecessary. There will always be at least one valid block in the ledger, the genesis block.

Reply:
remove is_valid_last_block